python.md

```diff
@@ -0,0 +1,1376 @@
+
+Your callback must handle at least two cases - a reason value of
+`rpm.RPMCALLBACK_INST_OPEN_FILE` and `rpm.RPMCALLBACK_INST_CLOSE_FILE.`
+
+With the reason of `rpm.RPMCALLBACK_INST_OPEN_FILE`, you must open the RPM
+package file and return a file descriptor for the file. You need to keep this
+file descriptor in a global-scope or otherwise-accessible variable, because with
+the reason of `rpm.RPMCALLBACK_INST_CLOSE_FILE`, you must close this file.
+
+
+## Coding A Sample Callback
+
+The following code shows a valid sample callback for upgrading and installing
+packages.
+
+```python
+# Global file descriptor for the callback.
+rpmtsCallback_fd = None
+
+def run_callback(reason, amount, total, key, client_data):
+    global rpmtsCallback_fd
+    if reason == rpm.RPMCALLBACK_INST_OPEN_FILE:
+        print("Opening file. ", reason, amount, total, key, client_data)
+        rpmtsCallback_fd = os.open(key, os.O_RDONLY)
+        return rpmtsCallback_fd
+    elif reason == rpm.RPMCALLBACK_INST_START:
+        print("Closing file. ", reason, amount, total, key, client_data)
+        os.close(rpmtsCallback_fd)
+```
+
+This callback assumes that the call to `addInstall` passed client data of the
+package file name. This callback ignores the `client_data` passed to the run
+method, but this is a perfect slot for passing an object. You can use this, for
+example, to avoid having a global variable for the file descriptor.
+
+
+## Upgrading A Package
+
+This example shows a simple Python script to upgrade or install a package.
+
+```python
+#!/usr/bin/python
+# Upgrades packages passed on the command line.
+# Usage:
+# python rpmupgrade.py rpm_file1.rpm rpm_file2.rpm ...
+#
+
+import os
+import sys
+import rpm
+
+# Global file descriptor for the callback.
+rpmtsCallback_fd = None
+
+
+def run_callback(reason, amount, total, key, client_data):
+    global rpmtsCallback_fd
+    if reason == rpm.RPMCALLBACK_INST_OPEN_FILE:
+        print("Opening file. ", reason, amount, total, key, client_data)
+        rpmtsCallback_fd = os.open(key, os.O_RDONLY)
+        return rpmtsCallback_fd
+    elif reason == rpm.RPMCALLBACK_INST_START:
+        print("Closing file. ", reason, amount, total, key, client_data)
+        os.close(rpmtsCallback_fd)
+
+
+def check_callback(ts, TagN, N, EVR, Flags):
+    if TagN == rpm.RPMTAG_REQUIRENAME:
+        prev = ""
+    Nh = None
+
+    if N[0] == "/":
+        dbitag = "basenames"
+    else:
+        dbitag = "providename"
+
+    # What do you need to do.
+    if EVR:
+        print("Must find package [", N, "-", EVR, "]")
+    else:
+        print("Must find file [", N, "]")
+
+    # FIXME
+    # if resolved:
+    #     # ts.addIntall(h, h, 'i')
+    #     return -1
+
+    return 1
+
+
+def read_rpm_header(ts, filename):
+    fd = os.open(filename, os.O_RDONLY)
+    try:
+        h = ts.hdrFromFdno(fd)
+    finally:
+        os.close(fd)
+    return h
+
+
+ts = rpm.TransactionSet()
+
+# Set to not verify DSA signatures.
```

Comment:
This does a whole lot more than just disable DSA - DSA which is no longer used anywhere.